### ISSUE TITLE
Some recipes show "undefined" ingredients in My Lego Network, and are not craftable

### ISSUE THREAD

**Issue by alyphen** [ruffle-rs/ruffle#14458]
### Describe the bug

![image](https://github.com/ruffle-rs/ruffle/assets/1935127/041e2f7b-abc2-4ce6-adf7-68a0ec5a2a96)
Acquire a recipe with 2 or more ingredients, such as the [Honey Pie](https://mylegonetwork.fandom.com/wiki/Honey_Pie).
Go to the private view, in the inventory, and select the recipe.
Recipe shows "undefined" instead of the top ingredient.

### Expected behavior

The first ingredient's name and required quantity are displayed, like all other ingredients.

### Content Location

https://mln.lcdruniverse.org/

### Affected platform

Browser's extension

### Operating system

Linux Mint 21.2

### Browser

Firefox 120.0.1

**Comment by n0samu** [comment]
Hi, this page requires login. Is there any way to access it?

**Comment by alyphen** [comment]
There's a sign-up link at the bottom of the page, which is a fairly uninvolved process, or it's possible to self-host too: https://github.com/MellonNet/mln-backend-emulator/
If there are any logs or diagnostics I can supply that would help I can do so

**Comment by n0samu** [comment]
Oh, the text at the bottom is nearly the same color as the background so I did not notice it 😅 
I've been able to reproduce this with the Crimson Apple Masterpiece blueprint that you get during the tutorial.

**Comment by Levi-Lesches** [comment]
Hi @n0samu, I'm one of the devs on the project. We basically re-use the flash files from the original Lego website, with very small modifications and with a rebuilt server backend in python. The bug does not reproduce on Basilisk, which makes me believe it's a ruffle issue. I'm happy to help debug and provide flash files if that would help

**Comment by n0samu** [comment]
Yep, I confirmed back then that it is a Ruffle issue. These kinds of games are very difficult to debug though and we aren't spending that much development time on AVM1 lately. I'll be sure to follow up with you a dev takes interest in working on a fix for this game. Thanks!

**Comment by Levi-Lesches** [comment]
Good to know! Also, don't waste _too_ much time on this because our long-term goal (that no one is actually working on) is porting the front end to a newer technology that will work on modern browsers

**Comment by Levi-Lesches** [comment]
Hi @n0samu, I was wondering if you could share more details on what the issue could have been, and what changes we might need to make to the Flash files to get them to play nicely with Ruffle. If it's not a lot, I may want to do that instead of rewriting the Flash bits. For reference, the project uses ActionScript 2, SWF version 9

**Comment by Levi-Lesches** [comment]
For reference, the project uses ActionScript 2, SWF version 9. Here is the relevant ActionScript: 

```actionscript
// If there's more than one required item
if (_loc4_.details.requirements.item._type == "array") {
   //  _loc3_ = all the required items as an array
   _loc3_ = ItemStep.theRoot.f_objToArray(_loc4_.details.requirements.item);
   _loc1_ = 1;
   while (_loc1_ < _loc3_.length) {
      // _loc10_ = the required item at index _loc1_.
      _loc10_ = _loc3_[_loc1_];
      _loc2_.needBoxCon.con.attachMovie("needItem","item" + _loc1_,_loc1_);
      _loc7_ = _loc2_.needBoxCon.con["item" + _loc1_];
      // _loc8_ = the amount of this item the user currently has
      _loc8_ = ItemStep.theRoot.o_inventoryPersonal[ItemStep.theRoot.o_itemHash[_loc10_.id].type][_loc10_.id].qty;
      if (!ItemStep.theRoot.o_inventoryPersonal[ItemStep.theRoot.o_itemHash[_loc10_.id].type][_loc10_.id]) {
         _loc8_ = 0;
      }
      // _loc7_ = the text element for this item
      _loc7_.qty.text = _loc8_ + "/" + _loc10_.qty;  // <-- reads: 0/undefined
      _loc7_.title.text = ItemStep.theRoot.o_itemHash[_loc10_.id].name;  // <-- reads: undefined
      _loc7_._y = ItemStep.settings.needItemHeight * (_loc1_ - 1);
      _loc1_ = _loc1_ + 1;
   }
}
```

And here is the relevant XML: 

```xml
<item name="Honey Pie Blueprint" type="blueprint" id="45120" thumbnail="/static/Blueprints/My LEGO Network/Thumbnails/Honey Pie Blueprint.png" branch="43041" description="Allows you to make a Honey Pie.">
   <details>
      <builds>
         <item qty="1" id="43575"/>
      </builds>
      <requirements>
         <item qty="1" id="43578"/>  <!-- 1 Honey Pot -->
         <item qty="5" id="43448"/>  <!-- 5 Apples -->
      </requirements>
   </details>
</item>
```

Note that the XML parsing code seems to reverse arrays, so this would actually be `[5 Apples, 1 Honey Pot]`. On Basilisk, it correctly renders like this: 

<img width="914" height="687" alt="Image" src="https://github.com/user-attachments/assets/f1bf5d5f-60d2-49a6-8c7c-700368310838" />

I'd have to look more thoroughly into the XML parsing code, but one thing that caught my eye is that this appears to be an off-by-one index issue: `_loc1_` is the index in the while loop, and it starts at 1, not 0. 

This part of the code subtracts 1:

```actionscript
_loc7_._y = ItemStep.settings.needItemHeight * (_loc1_ - 1);
```

but this does not:

```actionscript
// _loc10_ = the required item at index _loc1_
_loc10_ = _loc3_[_loc1_];
```

Given that it renders fine on Basilisk, I'm hesitant to call starting at 1 a bug in the ActionScript. But it would make sense that Ruffle is somehow having trouble with this: it can't find the first element (5 Apples -> undefined), and when it thinks it's on the second element, it's actually on the first (1 Honey Pot -> 5 Apples). 

I'll see what I can do in the way of debugging this, but do you have any insight here?

**Comment by Levi-Lesches** [comment]
I changed these two lines: 

```diff
- _loc1_ = 1;
- while (_loc1_ < _loc3_.length) {
+ _loc1_ = 0;
+ while (_loc1_ < _loc3_.length + 1) {
```

Basically, I forced the game to load one more requirement on each end. 

Basilisk rendered: 

- undefined
- Apple
- Honey Pot
- undefined

Ruffle rendered: 

- Honey Pot
- undefined
- Apple
- undefined

It's more likely then that the problem is in the XML parsing logic, which is a lot more complex. Would you have an idea what part of Ruffle could be affecting this?

**Comment by Xiphoseer** [comment]
@Levi-Lesches what's the implementation of that `f_objToArray`? Is that some actionscript standard I'm not finding or something custom from MLN? It looks like any "implementation-specific" behavior going from XML to the array would be in there.

**Comment by Levi-Lesches** [comment]
Here: 

```actionscript
f_objToArray = function(o_object)
{
   var a_array;
   var i_counter_0;
   a_array = new Array();
   for(i_counter_0 in o_object)
   {
      a_array.push(o_object[i_counter_0]);
   }
   a_array.reverse();
   return a_array;
};
```

It's part of this bigger file, which may help give more context

[MLN_utils.txt](https://github.com/user-attachments/files/22250508/MLN_utils.txt)

**Comment by Xiphoseer** [comment]
Interesting. Given the ` requirements.item._type == "array"`, this section from MDN about `for..in` with arrays seems relevant: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Loops_and_iteration#arrays

> Although it may be tempting to use this as a way to iterate over Array elements, the for...in statement will return the name of your user-defined properties in addition to the numeric indexes.
>
> Therefore, it is better to use a traditional for loop with a numeric index when iterating over arrays, because the for...in statement iterates over user-defined properties in addition to the array elements, if you modify the Array object (such as adding custom properties or methods).

Additionally, from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#description

> The traversal order, as of modern ECMAScript specification, is well-defined and consistent across implementations. Within each component of the prototype chain, all non-negative integer keys (those that can be array indices) will be traversed first in ascending order by value, then other string keys in ascending chronological order of property creation.

So the traversal order in AS2 with an array-like object from XML may be implementation defined and differ between basilisk and ruffle.

**Comment by Levi-Lesches** [comment]
Makes sense. So the test would be to basically find and replace every `for..in` over arrays with a regular for loop. I'll try that some time this week -- thanks!

**Comment by Xiphoseer** [comment]
For MLN probably yes. For AS2 I think it's actually documented to be in reverse index order, ie. `arr[1]`, `arr[0]` then other properties of the prototype chain, though `length` seems to be skipped as well.

https://web.archive.org/web/20090131183725/http://help.adobe.com/en_US/AS2LCR/Flash_10.0/help.html?content=Part2_AS2_LangRef_1.html

That matches the `reverse` the helper is doing. If MLN starts to read at 1, that may hint that arrays from XML have one other property, which ends up in front. Basilisk may be helpful to find out what that key is.

So i would say ruffle may need to "fix" the order for AS2.

**Comment by n0samu** [comment]
Iteration order is very hard to get right, so I'm not surprised that that may be the issue! Good to hear that in this particular case, the behavior is documented and consistent. Sounds like it should be fixable then! I've linked this thread in a dev channel to hopefully draw their attention and get a fix moving.

**Comment by Xiphoseer** [comment]
Looked a bit at where this may be implemented, and I think it's in `Action::Enumerate`, c.f. https://open-flash.github.io/documentation/avm1/actions/enumerate/. There's a very similar comment in https://github.com/ruffle-rs/ruffle/blob/28997d30d58a3ec559e0696b42770f6ff96992c2/core/src/avm1/object/script_object.rs#L747 which seems relevant.

If the `for..in` pops keys off the stack until there's an undefined, that's where the reverse comes in when the keys are pushed in index order. The proto keys coming first in that getter means they're last to show up in iteration. I'm not sure this already ensures that no key appears more than once. It may just be that arrays aren't considered there.

It's possible one could reduce the number of `Vec` allocations in that code by just iterating or recursing up/down the prototype chain while pushing values, keeping track of keys already seen.

Just wanted to mention this as a place to start debugging, entirely possible that the actual issue is somewhere else.

**Comment by Xiphoseer** [comment]
And yes, I was wondering how that code could use just attribute and element names to traverse XML, but I should just have looked at the helpers, it's not built-in.

It's the custom `XML.prototype.makeObj` and that also explains that the `_type` is not a weird decompiled `typeof` as I initially thought (which also never returns "array" anyway) so that's at index 0

**Comment by n0samu** [comment]
Sounds like you're getting somewhere with figuring it out - I'd be excited to see a PR from you! No pressure though 😄

**Comment by Xiphoseer** [comment]
Ok knowing AMF and seeing [`Array`](https://github.com/ruffle-rs/ruffle/blob/3fa35f6163600fce0e7a36f0bb436ed35e6318cb/core/src/avm1/globals/array.rs#L188) and [`PropertyMap`](https://github.com/ruffle-rs/ruffle/blob/3fa35f6163600fce0e7a36f0bb436ed35e6318cb/core/src/avm1/property_map.rs#L18) its possible that AVM was using some sort of dense array + sparse properties instead of just a single insertion order map.

Either way, insertion order matches that helper when
- 1st elem is set on the prop of its name
- 2nd elem reads _type
- wraps it in an array with one elem
- adds the `_type = "array"`
- then unshifts second object

But also this sounds familiar, but I don't have an explanation for the 2,0,1 order. https://stackoverflow.com/questions/10377042/flash-as2-for-each-loop-returning-odd-order-array

Still all theoretical though
